core/src/main/java/com/mapzen/android/core/MapzenManager.java

```diff
@@ -89,4 +97,23 @@ public void setApiKey(String apiKey) {
   public static String getSdkVersion() {
     return BuildConfig.SDK_VERSION;
   }
+
+  /**
+   * Adds listener to list of managed callbacks so that it can be notified when API key changes
+   * occur.
+   * @param listenerReference
+   */
+  public void addApiKeyChangeListener(WeakReference<ApiKeyChangeListener> listenerReference) {
+    Collections.synchronizedList(listeners).add(listenerReference);
+  }
+
+  private void notifyListeners() {
+    for (WeakReference<ApiKeyChangeListener> weakReference : Collections.synchronizedList(
```

Comment:
Just for my knowledge sake :D, can you explain the need for `WeakReference` here?

Reply:
The APIs aren't balanced with `addApiKeyChangeListener` / `removeApiKeyChangeListener` , there is only  `addApiKeyChangeListener` (this is because the `finalize` method isn't a reliable place to cleanup resources so I only have `addApiKeyChangeListener`). Because of this, I use `WeakReference` to avoid listeners outliving their containing objects and leaking resources when added.